### PR TITLE
PPLUS-105: Feature implement CLI commands

### PR DESCRIPTION
**Pull request by VladislavStrelchenko**
Overview
Ticket: https://spryker.atlassian.net/browse/PPLUS-105

Change log
Added upgrader upgrade command.
The command contains:

checks, if exist any uncommitted changes and return the error "Please commit or revert your changes" when exists.
run "composer update" command after checks.

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+    "require": {
+        "php": ">=7.3",
+        "ext-json": "*",
+        "ergebnis/json-printer": "^3.1",
+        "symfony/console": "^5.0",
+        "symfony/options-resolver": "^5.0",
+        "symfony/process": "^5.0"
+    },
+    "require-dev": {
+        "spryker/code-sniffer": "^0.15.6"
```
[the remark this answers is not in the document]
Comment (VladislavStrelchenko) [reply composer.json:18]: Tested https://github.com/spryker/code-sniffer with PhpStan locally 
Results: 
* level 6 -  [OK] No errors  
* level 8 -  [ERROR] Found 117 errors

[the remark this answers is not in the document]
Comment (dereuromark) [reply composer.json:18]: This repo, not code sniffer

File: .gitattributes [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+*.eot binary
+*.ico binary
+*.mo binary
+*.pdf binary
+*.xsd binary
+*.ts binary
+*.exe binary
+
+# Remove files for archives generated using `git archive`
+codeception.yml export-ignore
```
Comment (dereuromark) [inline .gitattributes:24]: This file should not be excluded, where did you get this from? We also need to fix up that tool

File: composer.json (hunk 1 above)
[the remark this answers is not in the document]
Comment (VladislavStrelchenko) [reply composer.json]: Added PhpStan https://github.com/spryker-sdk/upgrader/pull/2/commits/48c4b765eb822c4c0261564f0d74bfbfe9264353 
Also set level 5 for checks in GitHub CI https://github.com/spryker-sdk/upgrader/runs/3462043614?check_suite_focus=true

File: .gitattributes (hunk 2 above)
Comment (VladislavStrelchenko) [reply .gitattributes]: Removed https://github.com/spryker-sdk/upgrader/pull/2/commits/c9afde3077cc405f57fd460a5bcf7eb9edf8184c
Copied from there https://github.com/spryker-sdk/brancho/blob/master/.gitattributes

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+jobs:
+  validation:
+    runs-on: ubuntu-18.04
+    strategy:
+      fail-fast: false
+      matrix:
+        php-version: [
+            '7.3',
+            '7.4'
```
Comment (dereuromark) [inline .github/workflows/ci.yml:19]: should be 8.0 now
where did this template come from?

File: phpstan.neon [hunk 4]
```diff
@@ -0,0 +1,4 @@
+parameters:
+    level: 5
```
Comment (dereuromark) [inline phpstan.neon:2]: can we try 8?
all code, core and project should aim to that level

File: src/Upgrader/Business/Git/CommandResolver/UpdateIndexCommand.php [hunk 5]
```diff
@@ -28,6 +28,6 @@ public function isIndexOutdated(): bool
     {
         $commandResultDto = $this->exec($this->getCommand() . ' ' . self::REFRESH_FLAG);
 
-        return $commandResultDto->getResultCode();
+        return boolval($commandResultDto->getResultCode());
```
Comment (dereuromark) [inline src/Upgrader/Business/Git/CommandResolver/UpdateIndexCommand.php:31]: why is this needed? why not a (bool) cast instead?

Comment (VladislavStrelchenko) [reply src/Upgrader/Business/Git/CommandResolver/UpdateIndexCommand.php]: Fixed

File: phpstan.neon (hunk 4 above)
Comment (VladislavStrelchenko) [reply phpstan.neon:2]: Updated to level 8  
We have two types of PhpStan errors: 
 *  return type has no value type specified in iterable type Symfony\Component\Process\Process 
 * return type has no value type specified in iterable type array  

iterable type array - will fix in next story (when will implement valuable work with JSON files)

File: .github/workflows/ci.yml (hunk 3 above)
Comment (VladislavStrelchenko) [reply .github/workflows/ci.yml]: Added 8.0 
Template https://github.com/spryker/architecture-sniffer/blob/master/.github/workflows/ci.yml#L17

File: src/Upgrader/Business/Git/CommandResolver/UpdateIndexCommand.php [hunk 6]
```diff
@@ -28,6 +28,6 @@ public function isIndexOutdated(): bool
     {
         $commandResultDto = $this->exec($this->getCommand() . ' ' . self::REFRESH_FLAG);
 
-        return boolval($commandResultDto->getResultCode());
+        return (bool)$commandResultDto->getResultCode();
```
Comment (dereuromark) [inline src/Upgrader/Business/Git/CommandResolver/UpdateIndexCommand.php:31]: My question still stands. Why is this done here? Isnt the error code inverse to the boolean value?

Comment (VladislavStrelchenko) [reply src/Upgrader/Business/Git/CommandResolver/UpdateIndexCommand.php:31]: '(bool)' we need for avoid PhpStan error: Method Upgrader\Business\Git\CommandResolver\UpdateIndexCommand::isIndexOutdated() should return bool but returns int.

Comment (VladislavStrelchenko) [reply src/Upgrader/Business/Git/CommandResolver/UpdateIndexCommand.php:31]: If we have error code (1, 2 ,15 .... ) from getResultCode() it's mean `git update-index --refresh` failed as result we know that project has uncommitted changes

Comment (dereuromark) [reply src/Upgrader/Business/Git/CommandResolver/UpdateIndexCommand.php:31]: OK

File: phpstan.neon [hunk 7]
```diff
@@ -0,0 +1,4 @@
+parameters:
+    level: 8
+    paths:
+        - src/
```
Comment (dereuromark) [inline phpstan.neon:4]: iterable type can be ignored, see other modules

File: src/Upgrader/Business/CommandExecutor/AbstractCommandExecutor.php [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+
+namespace Upgrader\Business\CommandExecutor;
+
+use Exception;
+use Symfony\Component\Process\Process;
+
+abstract class AbstractCommandExecutor implements CommandExecutorInterface
+{
+    private const DEFAULT_COMMAND_TIMEOUT = 60;
+    private const ERROR_DESCRIPTION = 'Command: [%s]' . PHP_EOL . 'Error: %s';
```
Comment (dereuromark) [inline src/Upgrader/Business/CommandExecutor/AbstractCommandExecutor.php:16]: we never use private, only protected

File: src/Upgrader/Business/CommandExecutor/AbstractCommandExecutor.php [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+
+    /**
+     * @var \Upgrader\Business\CommandExecutor\CommandResultBuilder
+     */
+    private $resultBuilder;
```
Comment (dereuromark) [inline src/Upgrader/Business/CommandExecutor/AbstractCommandExecutor.php:21]: same

File: src/Upgrader/Business/CommandExecutor/AbstractCommandExecutor.php [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+     * @return \Upgrader\Business\CommandExecutor\CommandResultDto
+     */
+    public function execSuccess(?string $command = null): CommandResultDto
+    {
+        $process = $this->run($command);
+
+        if ($process->getExitCode()) {
+            $errorDescription = sprintf(self::ERROR_DESCRIPTION, $process->getCommandLine(), $process->getErrorOutput());
+
+            throw new Exception($errorDescription);
```
Comment (alexanderM91) [inline src/Upgrader/Business/CommandExecutor/AbstractCommandExecutor.php:55]: Please use custom exception.

File: src/Upgrader/Business/CommandExecutor/AbstractCommandExecutor.php [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+    }
+
+    /**
+     * @param string|null $command
+     *
+     * @throws \Exception
+     *
+     * @return \Upgrader\Business\CommandExecutor\CommandResultDto
+     */
+    public function execSuccess(?string $command = null): CommandResultDto
```
Comment (alexanderM91) [inline src/Upgrader/Business/CommandExecutor/AbstractCommandExecutor.php:48]: Why do we have a separate method?
Can we use one method exec instead for exec and execSuccess?

File: src/Upgrader/Business/CommandExecutor/CommandResultDto.php [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+ */
+
+namespace Upgrader\Business\CommandExecutor;
+
+class CommandResultDto
+{
+    /**
+     * @var int
+     */
+    private $resultCode;
```
Comment (alexanderM91) [inline src/Upgrader/Business/CommandExecutor/CommandResultDto.php:15]: Please use protected instead of private

File: src/Upgrader/Business/Composer/CommandExecutor/UpdateCommand.php [hunk 13]
```diff
@@ -0,0 +1,29 @@
...
+     */
+    public function getCommand(): string
+    {
+        return 'composer update';
+    }
+
+    /**
+     * @return int
+     */
+    public function getRequestTimeout(): int
```
Comment (alexanderM91) [inline src/Upgrader/Business/Composer/CommandExecutor/UpdateCommand.php:25]: Maybe it could be configurable option for all commands?

File: src/Upgrader/Business/Composer/ComposerJson/ComposerJsonReader.php [hunk 14]
```diff
@@ -0,0 +1,36 @@
...
+    {
+        $this->jsonFileReadHelper = $jsonFileReadHelper;
+    }
+
+    /**
+     * @return array
+     */
+    public function read(): array
+    {
+        return $this->jsonFileReadHelper->readFromPath(self::FILENAME_JSON);
```
Comment (alexanderM91) [inline src/Upgrader/Business/Composer/ComposerJson/ComposerJsonReader.php:34]: As for me helper is not required as it's just a proxy.
Please move the logic from Helper to the current Reader and the same think for Composer writer.

File: src/Upgrader/Business/Composer/ComposerJson/ComposerJsonWriter.php [hunk 15]
```diff
@@ -0,0 +1,38 @@
...
+use Upgrader\Business\Composer\Helper\JsonFile\JsonFileWriteHelper;
+
+class ComposerJsonWriter implements ComposerJsonWriterInterface
+{
+    protected const FILENAME_JSON = 'composer.json';
+
+    /**
+     * @var \Upgrader\Business\Composer\Helper\JsonFile\JsonFileWriteHelper
+     */
+    protected $JsonFileWriteHelper;
```
Comment (alexanderM91) [inline src/Upgrader/Business/Composer/ComposerJson/ComposerJsonWriter.php:19]: Json => json.

File: src/Upgrader/Business/Composer/ComposerJson/ComposerJsonWriter.php [hunk 16]
```diff
@@ -0,0 +1,38 @@
...
+    }
+
+    /**
+     * @param array $composerJsonArray
+     *
+     * @return bool
+     */
+    public function write(array $composerJsonArray): bool
+    {
+        return $this->JsonFileWriteHelper->writeToPath(self::FILENAME_JSON, $composerJsonArray);
```
Comment (alexanderM91) [inline src/Upgrader/Business/Composer/ComposerJson/ComposerJsonWriter.php:36]: Please check the comment related to Composer reader.

File: src/Upgrader/Business/Composer/Helper/JsonFile/JsonFileHelperFactory.php [hunk 17]
```diff
@@ -0,0 +1,50 @@
...
+            $this->jsonFileWriteHelper = new JsonFileWriteHelper($this->createJsonPrinter());
+        }
+
+        return $this->jsonFileWriteHelper;
+    }
+
+    /**
+     * @return \Ergebnis\Json\Printer\Printer
+     */
+    protected function createJsonPrinter(): Printer
```
Comment (alexanderM91) [inline src/Upgrader/Business/Composer/Helper/JsonFile/JsonFileHelperFactory.php:46]: Why do we need the current dependency?

File: src/Upgrader/Business/UpgraderFacade.php [hunk 18]
```diff
@@ -0,0 +1,72 @@
...
+
+    /**
+     * @param \Symfony\Component\Console\Input\InputInterface $input
+     * @param \Symfony\Component\Console\Output\OutputInterface $output
+     *
+     * @return int
+     */
+    public function upgrade(InputInterface $input, OutputInterface $output): int
+    {
+        $io = $this->createSymfonyStyle($input, $output);
```
Comment (alexanderM91) [inline src/Upgrader/Business/UpgraderFacade.php:32]: Can we move console command output into the AbstractCommand?

File: src/Upgrader/Business/CommandExecutor/CommandResultDto.php (hunk 12 above)
Comment (alexanderM91) [inline src/Upgrader/Business/CommandExecutor/CommandResultDto.php:15]: Please use protected instead of private

File: src/Upgrader/Business/UpgraderFacade.php [hunk 19]
```diff
@@ -0,0 +1,72 @@
...
+
+namespace Upgrader\Business;
+
+use Symfony\Component\Console\Input\InputInterface;
+use Symfony\Component\Console\Output\OutputInterface;
+use Symfony\Component\Console\Style\SymfonyStyle;
+
+class UpgraderFacade implements UpgraderFacadeInterface
+{
+    public const CODE_SUCCESS = 0;
```
Comment (alexanderM91) [inline src/Upgrader/Business/UpgraderFacade.php:16]: Can we move these constants into the AbstractCommand?

File: src/Upgrader/Business/UpgraderBusinessFactory.php [hunk 20]
```diff
@@ -0,0 +1,49 @@
...
+ * Use of this software requires acceptance of the Evaluation License Agreement. See LICENSE file.
+ */
+
+namespace Upgrader\Business;
+
+use Upgrader\Business\Composer\ComposerClient;
+use Upgrader\Business\Git\GitClient;
+use Upgrader\UpgraderConfig;
+
+class UpgraderBusinessFactory
```
Comment (alexanderM91) [inline src/Upgrader/Business/UpgraderBusinessFactory.php:14]: Can we have one factory that is responsible for objects initialisation?

File: src/Upgrader/Business/CommandExecutor/CommandResultDto.php [hunk 21]
```diff
@@ -0,0 +1,47 @@
+<?php
+
+/**
+ * Copyright © 2021-present Spryker Systems GmbH. All rights reserved.
+ * Use of this software requires acceptance of the Evaluation License Agreement. See LICENSE file.
+ */
+
+namespace Upgrader\Business\CommandExecutor;
+
+class CommandResultDto
```
Comment (alexanderM91) [inline src/Upgrader/Business/CommandExecutor/CommandResultDto.php:10]: Can we avoid DTO postfix and rename it e.g to CommandOutputResult or smth similar?

File: src/Upgrader/Business/Composer/CommandExecutor/UpdateCommand.php [hunk 22]
```diff
@@ -0,0 +1,29 @@
+<?php
+
+/**
+ * Copyright © 2021-present Spryker Systems GmbH. All rights reserved.
+ * Use of this software requires acceptance of the Evaluation License Agreement. See LICENSE file.
+ */
+
+namespace Upgrader\Business\Composer\CommandExecutor;
```
Comment (alexanderM91) [inline src/Upgrader/Business/Composer/CommandExecutor/UpdateCommand.php:8]: Are you sure that it's correct location, command executor, I think command is more suitable?

File: src/Upgrader/Business/Composer/ComposerClient.php [hunk 23]
```diff
@@ -0,0 +1,58 @@
...
+
+namespace Upgrader\Business\Composer;
+
+use Upgrader\Business\CommandExecutor\CommandResultDto;
+use Upgrader\Business\Composer\CommandExecutor\UpdateCommand;
+use Upgrader\Business\Composer\ComposerJson\ComposerJsonReader;
+use Upgrader\Business\Composer\ComposerLock\ComposerLockReader;
+use Upgrader\UpgraderConfig;
+
+class ComposerClient
```
Comment (alexanderM91) [inline src/Upgrader/Business/Composer/ComposerClient.php:16]: Please provide an abstraction.

**Review by alexanderM91 — CHANGES_REQUESTED** [review]

File: phpstan.neon (hunk 7 above)
Comment (VladislavStrelchenko) [reply phpstan.neon:4]: Added checkMissingIterableValueType: false to phpstan.neon

File: src/Upgrader/Business/CommandExecutor/AbstractCommandExecutor.php (hunk 8 above)
Comment (VladislavStrelchenko) [reply src/Upgrader/Business/CommandExecutor/AbstractCommandExecutor.php]: ++

File: src/Upgrader/Business/CommandExecutor/AbstractCommandExecutor.php (hunk 9 above)
Comment (VladislavStrelchenko) [reply src/Upgrader/Business/CommandExecutor/AbstractCommandExecutor.php]: ++

File: src/Upgrader/Business/CommandExecutor/AbstractCommandExecutor.php (hunk 10 above)
Comment (VladislavStrelchenko) [reply src/Upgrader/Business/CommandExecutor/AbstractCommandExecutor.php]: Created UpgraderCommandExecException

File: src/Upgrader/Business/CommandExecutor/AbstractCommandExecutor.php (hunk 11 above)
Comment (VladislavStrelchenko) [reply src/Upgrader/Business/CommandExecutor/AbstractCommandExecutor.php]: Removed this method, still is only "exec"

File: src/Upgrader/Business/CommandExecutor/CommandResultDto.php (hunk 12 above)
Comment (VladislavStrelchenko) [reply src/Upgrader/Business/CommandExecutor/CommandResultDto.php]: ++

File: src/Upgrader/Business/Composer/CommandExecutor/UpdateCommand.php (hunk 13 above)
Comment (VladislavStrelchenko) [reply src/Upgrader/Business/Composer/CommandExecutor/UpdateCommand.php]: Moved config to UpgraderConfig

File: src/Upgrader/Business/Composer/ComposerJson/ComposerJsonWriter.php (hunk 15 above)
Comment (VladislavStrelchenko) [reply src/Upgrader/Business/Composer/ComposerJson/ComposerJsonWriter.php]: ++

File: src/Upgrader/Business/Composer/ComposerJson/ComposerJsonWriter.php (hunk 16 above)
Comment (VladislavStrelchenko) [reply src/Upgrader/Business/Composer/ComposerJson/ComposerJsonWriter.php]: ++

File: src/Upgrader/Business/Composer/Helper/JsonFile/JsonFileHelperFactory.php (hunk 17 above)
Comment (VladislavStrelchenko) [reply src/Upgrader/Business/Composer/Helper/JsonFile/JsonFileHelperFactory.php]: For adjusting indentation to JSON string (AbstractJsonWriter.php)

File: src/Upgrader/Business/UpgraderFacade.php (hunk 18 above)
Comment (VladislavStrelchenko) [reply src/Upgrader/Business/UpgraderFacade.php]: ++

File: src/Upgrader/Business/CommandExecutor/CommandResultDto.php (hunk 12 above)
Comment (VladislavStrelchenko) [reply src/Upgrader/Business/CommandExecutor/CommandResultDto.php]: ++

File: src/Upgrader/Business/UpgraderFacade.php (hunk 19 above)
Comment (VladislavStrelchenko) [reply src/Upgrader/Business/UpgraderFacade.php]: ++

File: src/Upgrader/Business/UpgraderBusinessFactory.php (hunk 20 above)
Comment (VladislavStrelchenko) [reply src/Upgrader/Business/UpgraderBusinessFactory.php:14]: We have one factory on the Business layer and one factory for each client (ComposerClient, GitClinet, etc....)

File: src/Upgrader/Business/CommandExecutor/CommandResultDto.php (hunk 21 above)
Comment (VladislavStrelchenko) [reply src/Upgrader/Business/CommandExecutor/CommandResultDto.php]: Renamed to CommandResult

File: src/Upgrader/Business/Composer/CommandExecutor/UpdateCommand.php (hunk 22 above)
Comment (VladislavStrelchenko) [reply src/Upgrader/Business/Composer/CommandExecutor/UpdateCommand.php]: Renamed namespace

File: src/Upgrader/Business/Composer/ComposerClient.php (hunk 23 above)
Comment (VladislavStrelchenko) [reply src/Upgrader/Business/Composer/ComposerClient.php]: Created ComposerClientInterface

File: src/Upgrader/Business/PackageManager/Client/Composer/Json/Writer/ComposerJsonWriter.php [hunk 24]
```diff
@@ -0,0 +1,101 @@
...
+    /**
+     * @param string $path
+     * @param array $body
+     *
+     * @return bool
+     */
+    protected function writeToPath(string $path, array $body): bool
+    {
+        $indentation = $this->detectIndentation($path);
+        // ToDo Move to helper
```
Comment (alexanderM91) [inline src/Upgrader/Business/PackageManager/Client/Composer/Json/Writer/ComposerJsonWriter.php:52]: ToDo?

File: src/Upgrader/Business/UpgraderFacadeInterface.php [hunk 25]
```diff
@@ -0,0 +1,24 @@
...
+
+namespace Upgrader\Business;
+
+use Upgrader\Business\Command\ResultOutput\CommandResultOutput;
+
+interface UpgraderFacadeInterface
+{
+    /**
+     * Specification:
+     * - Updates Spryker packages.
```
Comment (alexanderM91) [inline src/Upgrader/Business/UpgraderFacadeInterface.php:16]: Could you please adjust specification

- Checks for uncommented changes.
- Updates packages to the latest available version.
- Returns command result object that contains exit code and message.

**Review by alexanderM91 — APPROVED** [review]

File: .github/workflows/ci.yml [hunk 26]
```diff
@@ -0,0 +1,61 @@
...
+
+jobs:
+  validation:
+    runs-on: ubuntu-18.04
+    strategy:
+      fail-fast: false
+      matrix:
+        php-version: [
+            '7.3',
+            '7.4',
```
Comment (dereuromark) [inline .github/workflows/ci.yml:18]: only min/max needed

**Review by dereuromark — APPROVED** [review]

**Comment by dereuromark** [comment]
Please remove merged branches afterwards
